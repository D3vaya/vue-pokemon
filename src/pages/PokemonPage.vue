<template>
  <h1 v-if="!pokemon">Espero por favor</h1>
  <div v-else>
    <h1>¿Quién es este pokémon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOption :pokemons="pokemonsArr" @selection="checkAnswer($event)" />
    <template v-if="showAnser">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">Nuevo Juego</button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOption from "@/components/PokemonOption.vue";
import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
  components: {
    PokemonPicture,
    PokemonOption,
  },
  data() {
    return {
      message: "",
      pokemon: null,
      pokemonsArr: [],
      showAnser: false,
      showPokemon: false,
    };
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonsArr = await getPokemonOptions();
      const rnd = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonsArr[rnd];
    },
    checkAnswer(selectedPokemonId) {
      this.showPokemon = true;
      this.showAnser = true;
      if (this.pokemon.id === selectedPokemonId) {
        this.message = `Correcto, es ${this.pokemon.name}`;
      } else {
        this.message = `Ooops, era ${this.pokemon.name}`;
      }
    },
    newGame() {
      this.pokemon = null;
      this.showPokemon = false;
      this.showAnser = false;
      this.pokemonsArr = [];
      this.mixPokemonArray();
    },
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<style></style>
