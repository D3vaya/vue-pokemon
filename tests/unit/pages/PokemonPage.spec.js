import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage";
import { pokemonsMock } from "../mocks/pokemons.mock";

describe("PokemonPage component", () => {
  let wrapper;
  //let mixPokemonArraySpy;
  beforeEach(() => {
    //mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, "mixPokemonArray");
    wrapper = shallowMount(PokemonPage);
  });
  test("should render PokemonPicture component", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Debe de llamar mixPokemonArray al montar", () => {
    const mixPokemonArraySpy = jest.spyOn(
      PokemonPage.methods,
      "mixPokemonArray"
    );
    const wrapper = shallowMount(PokemonPage);
    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  test("debe de hacer match con el snapshot cuando cargan los pokemons", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          message: "",
          showAnser: false,
          showPokemon: false,
          pokemon: pokemonsMock[0],
          pokemonsArr: pokemonsMock,
        };
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("debe de mostrar los componentes PokemonPicture y PokemonOption", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          message: "",
          showAnser: false,
          showPokemon: false,
          pokemon: pokemonsMock[0],
          pokemonsArr: pokemonsMock,
        };
      },
    });
    const picture = wrapper.find("pokemon-picture-stub");
    const option = wrapper.find("pokemon-option-stub");
    // que existan los componentes PokemonPicture y PokemonOption
    expect(picture.exists()).toBe(true);
    expect(option.exists()).toBe(true);
    // que tenga el atributo pokemonId en 5
    expect(wrapper.find("pokemon-picture-stub").attributes("pokemonid")).toBe(
      "3"
    );
    // que existe el atributo pokemons
    expect(
      wrapper.find("pokemon-option-stub").attributes("pokemons")
    ).toBeTruthy();
  });

  test("pruebas de checkAnswer", async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          message: "",
          showAnser: false,
          showPokemon: false,
          pokemon: pokemonsMock[0],
          pokemonsArr: pokemonsMock,
        };
      },
    });
    await wrapper.vm.checkAnswer(3);
    expect(wrapper.find("h2").exists()).toBe(true);
    expect(wrapper.vm.showPokemon).toBe(true);
    expect(wrapper.find("h2").text()).toBe(
      `Correcto, es ${pokemonsMock[0].name}`
    );
    await wrapper.vm.checkAnswer(100);
    expect(wrapper.vm.message).toBe(`Ooops, era ${pokemonsMock[0].name}`);
  });
});
