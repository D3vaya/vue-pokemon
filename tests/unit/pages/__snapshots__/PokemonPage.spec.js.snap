// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokémon?</h1>
  <pokemon-picture-stub pokemonid="3" showpokemon="false"></pokemon-picture-stub>
  <pokemon-option-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-option-stub>
  <!--v-if-->
</div>
`;

exports[`PokemonPage component should render PokemonPicture component 1`] = `<h1>Espero por favor</h1>`;
